python/ray/_private/runtime_env.py

```diff
@@ -40,11 +40,17 @@ class RuntimeEnvDict:
             modules to add to the `sys.path`.
             Examples:
                 ["/path/to/other_module", "/other_path/local_project.zip"]
+        pip (str): A string containing the contents of a pip requirements.txt
+            file.  This will be dynamically installed in the runtime env.
```

Comment:
This may be a dumb question, but is there a reason this is a str, not a dict?

Reply:
No, I don't think there's any particular reason.  @tchordia is it okay if we just make the "pip" field a `List[str]` representing a list of pip packages, instead of a `str` containing `requirements.txt`?  It'll simplify the implementation on the backend and it seems easier to specify in Python, but I remember you had some insight about this

Reply:
Yeah, we had this issue on the product. Most users start with a requirements.txt file, not a list of strings. If you accept a list of pip packages in your API, most users will need to translate their requirements.txt into a list of pip packages. This is not necessarily trivial, as requirements.txt files can have a bunch of fancy things (comments, line continuations etc). It seems like we can save the user this work by just accepting a requirements.txt file.

What is the advantage to having a list of strings? I think I'm not really following that bit.

Reply:
That makes sense to me--so @edoakes the only additional pip-specific code on the backend would be to save this requirements.txt on disk so it can be installed, is that okay?  And then to automatically insert the Ray dependency, I can just insert it as the first line of the requirements.txt file.

Reply:
This is in-code, writing/pasting a whole requirements.txt spec into python code is going to be really ugly. Imagine a simple case where I just want pkg1 and pkg2:

```python
ray.init(env={"pip": "pkg1\npkg2\n"})
```
or
```python
pip_env = """
pkg1
pkg2
"""
ray.init(env={"pip": pip_env})
```

At the very least we also need to support a list of strings for simple cases like this so I'd prefer to start with just that

Reply:
Ahh, that makes sense @tchordia !! I think supporting both is quite trivial because going from `List[str]` -> `str` is just `"\n".join(list)`.

Reply:
Great idea @ijrsvt, I think that addresses @edoakes concern as well as @tchordia's.  So I'll move ahead with supporting both `List[str]`and `str`in the user-facing API, and in the backend it will be stored as a `str` in the runtime env.  In the worker nodes, I'll save the `str` to a temporary file `requirements.txt` and include `-r requirements.txt` in the conda spec when installing.